Populates no_proxy to all hosts

cbayle opened the pull request:
/kind bug

Fixes #6022 

```release-note
Fix deployment failure when behind a proxy
```

floryut commented:
Hello @cbayleOL 
Thanks for the PR, but can you take a look at @Miouge1 fix which is merged in master a few minutes ago ? #6039

It should fix your bug and proxy related ones.

cbayle commented:
Unfortunatly, it doesn't fix the issue, but create a new one in :
```
TASK [container-engine/docker : Ensure old versions of Docker are not installed. | Debian] ***
ok: [k8s-2]
ok: [k8s-1]
ok: [k8s-3]
Thursday 30 April 2020  15:18:27 +0200 (0:00:00.651)       0:02:20.774 ******** 
Thursday 30 April 2020  15:18:27 +0200 (0:00:00.046)       0:02:20.820 ******** 
FAILED - RETRYING: ensure docker-ce repository public key is installed (4 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (4 retries left).

```

Nevertheless, may path is breaking the automated build :-(

Miouge1 commented:
@cbayle in #6039 I'm doing almost exactly the same (in the kubespray-defaults role rather than kubernetes-apps/ansible role). I also added a test job to verify that it worked.

If there is still a problem on latest master (commit 353d44a4a6abb3be23b6af9096b75e1160bd9e92), then I invite you send info on how to reproduce it locally with Vagrant.

cbayle commented:
@Miouge1 I committed a Makefile in my branch cbayle/kubespray:work-in-progress 
https://raw.githubusercontent.com/cbayle/kubespray/work-in-progress/Makefile

You have to set the PROXY in the Makefile and if you have vagrant installed 'make' will pop
 a k8s cluster 

In the container registry setup, you should see it is not working for docker.
with ubuntu this is because of netplan config that tries to lookup google servers
I made some change in roles/bootstrap-os/tasks/bootstrap-debian.yml to fix this
https://github.com/cbayle/kubespray/commit/a586354860495dd01d677ceb06d1ae83e3489f7a#diff-fbdd08c0a1cb3e9d3cc217c33b2d5b61

You should get this

```
TASK [container-engine/docker : Ensure old versions of Docker are not installed. | Debian] ***
ok: [k8s-1]
ok: [k8s-2]
ok: [k8s-3]
Thursday 30 April 2020  19:39:03 +0200 (0:00:00.657)       0:02:19.775 ******** 
Thursday 30 April 2020  19:39:03 +0200 (0:00:00.058)       0:02:19.833 ******** 
FAILED - RETRYING: ensure docker-ce repository public key is installed (4 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (4 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (4 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (3 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (3 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (3 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (2 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (2 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (2 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (1 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (1 retries left).
FAILED - RETRYING: ensure docker-ce repository public key is installed (1 retries left).

TASK [container-engine/docker : ensure docker-ce repository public key is installed] ***
failed: [k8s-2] (item=9DC858229FC7DD38854AE2D88D81803C0EBFCD88) => {"ansible_loop_var": "item", "attempts": 4, "changed": false, "item": "9DC858229FC7DD38854AE2D88D81803C0EBFCD88", "msg": "Failed to download key at https://download.docker.com/linux/ubuntu/gpg: Request failed: <urlopen error [Errno 99] Cannot assign requested address>"}
failed: [k8s-3] (item=9DC858229FC7DD38854AE2D88D81803C0EBFCD88) => {"ansible_loop_var": "item", "attempts": 4, "changed": false, "item": "9DC858229FC7DD38854AE2D88D81803C0EBFCD88", "msg": "Failed to download key at https://download.docker.com/linux/ubuntu/gpg: Request failed: <urlopen error [Errno 99] Cannot assign requested address>"}
failed: [k8s-1] (item=9DC858229FC7DD38854AE2D88D81803C0EBFCD88) => {"ansible_loop_var": "item", "attempts": 4, "changed": false, "item": "9DC858229FC7DD38854AE2D88D81803C0EBFCD88", "msg": "Failed to download key at https://download.docker.com/linux/ubuntu/gpg: Request failed: <urlopen error [Errno 99] Cannot assign requested address>"}

```

That shows some trouble with proxy, I you comment  the lines you added in roles/kubespray-defaults/tasks/no_proxy.yml
```
#    proxy_env:
#      no_proxy: "{{ hostvars.localhost.no_proxy_prepare }}"
#      NO_PROXY: "{{ hostvars.localhost.no_proxy_prepare }}"

```

run 'make ansible', you should get this : 
```
TASK [container-engine/docker : ensure docker-ce repository public key is installed] *****************************************************************************************************
changed: [k8s-1] => (item=9DC858229FC7DD38854AE2D88D81803C0EBFCD88)
changed: [k8s-3] => (item=9DC858229FC7DD38854AE2D88D81803C0EBFCD88)
changed: [k8s-2] => (item=9DC858229FC7DD38854AE2D88D81803C0EBFCD88)

```

fixing the problem, quite strange indeed, but later the trouble occurs in 

```
TASK [kubespray-defaults : Configure defaults] *******************************************************************************************************************************************
ok: [k8s-1] => {
    "msg": "Check roles/kubespray-defaults/defaults/main.yml"
}
ok: [k8s-2] => {
    "msg": "Check roles/kubespray-defaults/defaults/main.yml"
}
Thursday 30 April 2020  19:55:33 +0200 (0:00:00.034)       0:09:52.863 ******** 
Thursday 30 April 2020  19:55:33 +0200 (0:00:00.026)       0:09:52.890 ******** 
Thursday 30 April 2020  19:55:33 +0200 (0:00:00.029)       0:09:52.920 ******** 
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (20 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (19 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (18 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (17 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (16 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (15 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (14 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (13 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (12 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (11 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (10 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (9 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (8 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (7 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (6 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (5 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (4 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (3 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (2 retries left).
FAILED - RETRYING: Kubernetes Apps | Wait for kube-apiserver (1 retries left).

TASK [kubernetes-apps/ansible : Kubernetes Apps | Wait for kube-apiserver] ***************************************************************************************************************
fatal: [k8s-1]: FAILED! => {"attempts": 20, "changed": false, "content": "", "elapsed": 0, "msg": "Status code was -1 and not [200]: Request failed: <urlopen error Tunnel connection failed: 403 Forbidden>", "redirected": false, "status": -1, "url": "https://127.0.0.1:6443/healthz"}

```

that is fixed by adding 

```
- name: Populates no_proxy to all hosts
  set_fact:
    proxy_env:
      no_proxy: "{{ hostvars.localhost.no_proxy_prepare }}"
```

at the beginning of roles/kubernetes-apps/ansible/tasks/main.yml
